lib/Conversion/ImportVerilog/Structure.cpp

```diff
@@ -0,0 +1,158 @@
+//===- Structure.cpp - Slang hierarchy conversion -------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "ImportVerilogInternals.h"
+#include "slang/ast/Compilation.h"
+
+using namespace circt;
+using namespace ImportVerilog;
+
+//===----------------------------------------------------------------------===//
+// Module Member Conversion
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct MemberVisitor {
+  Context &context;
+  Location loc;
+  OpBuilder &builder;
+
+  /// Skip semicolons.
+  LogicalResult visit(const slang::ast::EmptyMemberSymbol &) {
+    return success();
+  }
+
+  LogicalResult visit(const slang::ast::InstanceSymbol &instNode) {
+    auto targetModule = context.convertModuleHeader(&instNode.body);
+    if (!targetModule)
+      return failure();
+
+    builder.create<moore::InstanceOp>(
+        loc, builder.getStringAttr(instNode.name),
+        FlatSymbolRefAttr::get(targetModule.getSymNameAttr()));
+
+    return success();
+  }
+
+  /// Emit an error for all other members.
+  template <typename T>
+  LogicalResult visit(T &&node) {
+    mlir::emitError(loc, "unsupported construct: ")
+        << slang::ast::toString(node.kind);
+    return failure();
+  }
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Structure and Hierarchy Conversion
+//===----------------------------------------------------------------------===//
+
+/// Convert an entire Slang compilation to MLIR ops. This is the main entry
+/// point for the conversion.
+LogicalResult
+Context::convertCompilation(slang::ast::Compilation &compilation) {
+  const auto &root = compilation.getRoot();
+
+  // Visit all top-level declarations in all compilation units. This includes
+  // instantiable constructs like modules, interfaces, and programs, which are
+  // listed separately as top instances.
+  for (auto *unit : root.compilationUnits) {
+    for (const auto &member : unit->members()) {
+      // Error out on all top-level declarations.
+      auto loc = convertLocation(member.location);
+      return mlir::emitError(loc, "unsupported construct: ")
+             << slang::ast::toString(member.kind);
+    }
+  }
+
+  // Prime the root definition worklist by adding all the top-level modules.
+  SmallVector<const slang::ast::InstanceSymbol *> topInstances;
+  for (auto *inst : root.topInstances)
+    convertModuleHeader(&inst->body);
+
+  // Convert all the root module definitions.
+  while (!moduleWorklist.empty()) {
+    auto *module = moduleWorklist.front();
+    moduleWorklist.pop();
+    if (failed(convertModuleBody(module)))
+      return failure();
+  }
```

Comment:
I usually try to keep the number of class fields at a minimum to reduce the overall surface of side-effects a method can have, and here, the additional field didn't seem to be worth it just to save one additional function argument. However, since modules can be hidden in various places, it makes sense to me.